### PR TITLE
api logging proposal

### PR DESCRIPTION
**Pull request by bradymiller**
work in progress

@sjpadgett , What are your thoughts on this mechanism? Just trying to get something in place that essentially logs all api stuff.

### CODE REVIEW

File: _rest_config.php [hunk 1]
```diff
@@ -254,6 +254,11 @@ static function authentication_check($resource)
             }
         }
     }
+
+    static function apiLog($requestUrl, $response, $requestBody = '')
+    {
+        sqlStatementNoLog("INSERT INTO `api_log` (`user_id`, `patient_id`, `request_url`, `request_body`, `response`, `created_time`) VALUES (?, ?, ?, ?, ?, NOW())", [($_SESSION['authUserID'] ?? 0), ($_SESSION['pid'] ?? 0), $requestUrl, $requestBody, $response]);
```
Comment (sjpadgett) [inline _rest_config.php:260]: I'd add IP for phishers.
request bodys can get quite large. Maybe request type i.e login, auth, request action. Unsure what we gain with logging unless there's some intelligence built in to do something with the log or debugging.  Still it's a little closer to a review process which I still feel strongly about.

**Comment by sjpadgett** [comment]
okay by me Brady. Though logs make up 10-15% of backup size. Maybe more for those that don't manage them so, logging is always a concern for site admins.

File: _rest_routes.inc.php [hunk 2]
```diff
@@ -47,359 +47,528 @@
...
         RestConfig::authorization_check("encounters", "notes");
         $data = (array) (json_decode(file_get_contents("php://input")));
-        return (new EncounterRestController())->putVital($pid, $eid, $vid, $data);
+        $return = (new EncounterRestController())->putVital($pid, $eid, $vid, $data);
+        RestConfig::apiLog($return, $data);
+        return $return;
     },
     "GET /api/patient/:pid/encounter/:eid/vital" => function ($pid, $eid) {
         RestConfig::authorization_check("encounters", "notes");
         return (new EncounterRestController())->getVitals($pid, $eid);
```
Comment (bradymiller) [inline _rest_routes.inc.php:155]: oops

**Comment by bradymiller** [comment]
whooohooo, this is testing well and logs everything i could think of. Now have option of no api logging, minimal api logging, and full api logging. Also incorporates encryption when audit log encryption is turned on. Note it will not log internal api requests either.

**Comment by bradymiller** [comment]
travis is displaying passive-aggressive behavior (php8 nightly build is broken, which is likely temporary; will move that test to the allowed failure category for now until travis gets over its temper tantrum :)

**Comment by sjpadgett** [comment]
cool, i'm not the only one he picks on!!!

**Comment by bradymiller** [comment]
whooohooo, nothing like a travis takedown:

tenor (35)

bringing this PR in

**Pull request merged**